File: FBSimulatorControl/Commands/FBSimulatorCrashLogCommands.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBSimulatorCrashLogCommands.h"

#import "FBSimulator.h"

@interface FBSimulatorCrashLogCommands ()

@property (nonatomic, weak, readonly) FBSimulator *simulator;
@property (nonatomic, strong, readonly) FBCrashLogNotifier *notifier;
@property (nonatomic, assign, readwrite) BOOL hasPerformedInitialIngestion;

@end

@implementation FBSimulatorCrashLogCommands

#pragma mark Initializers

+ (instancetype)commandsWithTarget:(id<FBiOSTarget>)target
{
  NSParameterAssert([target isKindOfClass:FBSimulator.class]);
  return [[self alloc] initWithSimulator:(FBSimulator *)target notifier:FBCrashLogNotifier.sharedInstance];
}

- (instancetype)initWithSimulator:(FBSimulator *)simulator notifier:(FBCrashLogNotifier *)notifier
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _simulator = simulator;
  _notifier = notifier;
  _hasPerformedInitialIngestion = NO;

  return self;
}

#pragma mark id<FBiOSTarget>

- (FBFuture<FBCrashLogInfo *> *)notifyOfCrash:(NSPredicate *)predicate
{
  return [self.notifier nextCrashLogForPredicate:predicate];
}

- (FBFuture<NSArray<FBCrashLogInfo *> *> *)crashes:(NSPredicate *)predicate useCache:(BOOL)useCache
{
  if (!self.hasPerformedInitialIngestion) {
    [self.notifier.store ingestAllExistingInDirectory];
    self.hasPerformedInitialIngestion = YES;
  }

  return [FBFuture futureWithResult:[self.notifier.store ingestedCrashLogsMatchingPredicate:predicate]];
}

- (FBFuture<NSArray<FBCrashLogInfo *> *> *)pruneCrashes:(NSPredicate *)predicate
{
  // Unfortunately, the Crash Logs that are created for Simulators may not contain the UDID of the Simulator.
  // Crashes will not contain a UDID if they are launching System Apps that are present in the RuntimeRoot, not the Simulator Data Directory.
  // If they are Applications installed by the User, the UDID will appear in the launch path, as the Application is installed relative to the Simulator's Simulator Data Directory.
  // For this reason, we need to be conservative about which Crash Logs to prune, otherwise we may end up deleting the crash logs of another Simulator, or something running on the host.
  // Deleting these behind the back of the API is not something that makes sense.
  // We ensure that *any* crash logs that are to be deleted *must* contain the UDID of the Simulator.
  NSPredicate *simulatorPredicate = [NSCompoundPredicate andPredicateWithSubpredicates:@[
    [FBCrashLogInfo predicateForExecutablePathContains:self.simulator.udid],
    predicate,
  ]];
  return [FBFuture futureWithResult:[self.notifier.store pruneCrashLogsMatchingPredicate:simulatorPredicate]];
}

- (FBFutureContext<id<FBFileContainer>> *)crashLogFiles
{
  return [[FBControlCoreError
    describeFormat:@"%@ not supported on simulators", NSStringFromSelector(_cmd)]
    failFutureContext];
}

@end

File: FBSimulatorControl/Commands/FBSimulatorCrashLogCommands.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import FBControlCore
import Foundation

/// The way crash-log file access fails on simulators, as data rather than an assembled string.
public enum FBSimulatorCrashLogError: Error, LocalizedError {
  case fileAccessUnsupported

  public var errorDescription: String? {
    switch self {
    case .fileAccessUnsupported:
      return "crashLogFiles not supported on simulators"
    }
  }
}

public final class FBSimulatorCrashLogCommands {

  // MARK: - Properties

  private weak var simulator: FBSimulator?
  private let notifier: FBCrashLogNotifier
  private var hasPerformedInitialIngestion: Bool = false

  // MARK: - Initializers

  public class func commands(with simulator: FBSimulator) -> FBSimulatorCrashLogCommands {
    FBSimulatorCrashLogCommands(
      simulator: simulator,
      notifier: FBCrashLogNotifier.sharedInstance
    )
  }

  private init(simulator: FBSimulator, notifier: FBCrashLogNotifier) {
    self.simulator = simulator
    self.notifier = notifier
  }

  // MARK: - Private

  fileprivate func notifyOfCrashAsync(matching predicate: NSPredicate) async throws -> FBCrashLogInfo {
    try await notifier.nextCrashLog(forPredicate: predicate)
  }

  fileprivate func crashesAsync(matching predicate: NSPredicate, useCache: Bool) async throws -> [FBCrashLogInfo] {
    if !hasPerformedInitialIngestion {
      notifier.store.ingestAllExistingInDirectory()
      hasPerformedInitialIngestion = true
    }
    return notifier.store.ingestedCrashLogs(matchingPredicate: predicate)
  }

  fileprivate func pruneCrashesAsync(matching predicate: NSPredicate) async throws -> [FBCrashLogInfo] {
    guard let simulator = self.simulator else {
      throw FBWeakTargetError.simulator
    }
    let simulatorPredicate = NSCompoundPredicate(andPredicateWithSubpredicates: [
      FBCrashLogInfo.predicate(forExecutablePathContains: simulator.udid),
      predicate,
    ])
    return notifier.store.pruneCrashLogs(matchingPredicate: simulatorPredicate)
  }
}

// MARK: - FBSimulator+CrashLogCommands

extension FBSimulator: CrashLogCommands {

  public func crashes(matching predicate: NSPredicate, useCache: Bool) async throws -> [FBCrashLogInfo] {
    try await crashLogCommands().crashesAsync(matching: predicate, useCache: useCache)
  }

  public func notifyOfCrash(matching predicate: NSPredicate) async throws -> FBCrashLogInfo {
    try await crashLogCommands().notifyOfCrashAsync(matching: predicate)
  }

  public func pruneCrashes(matching predicate: NSPredicate) async throws -> [FBCrashLogInfo] {
    try await crashLogCommands().pruneCrashesAsync(matching: predicate)
  }

  public func withCrashLogFiles<R>(body: (any AsyncFileContainer) async throws -> R) async throws -> R {
    throw FBSimulatorCrashLogError.fileAccessUnsupported
  }
}
